File: 1335. Minimum Difficulty of a Job Schedule.java
class Solution {
  public int minDifficulty(int[] jobDifficulty, int d) {
    final int x = jobDifficulty.length;
    if (x < d)
      return -1;
    int[][] dp = new int[x + 1][d + 1];
    Arrays.stream(dp).forEach(row -> Arrays.fill(row, Integer.MAX_VALUE / 2));
    dp[0][0] = 0;

    for (int i = 1; i <= x; ++i)
      for (int k = 1; k <= d; ++k) {
        int maxDifficulty = 0;
        for (int j = i - 1; j >= k - 1; --j) {                       
          maxDifficulty = Math.max(maxDifficulty, jobDifficulty[j]); 
          dp[i][k] = Math.min(dp[i][k], dp[j][k - 1] + maxDifficulty);
        }
      }
    return dp[x][d];
  }
}

File: 1335. Minimum Difficulty of a Job Schedule.cpp
class Solution {
public:
    int minDifficulty(vector<int>& jobDifficulty, int d) {
        int x = jobDifficulty.size();
     int f[x+1][d+1];
      memset(f, 0x3f, sizeof(f));
      f[0][0] = 0;
       for (int i = 1; i <= x; ++i) {
            for (int j = 1; j <= min(d, i); ++j) {
                int mx = 0;
                for (int k = i; k; --k) {
                    mx = max(mx, jobDifficulty[k - 1]);
                    f[i][j] = min(f[i][j], f[k - 1][j - 1] + mx);
                }
            }
        }
        return f[x][d] == 0x3f3f3f3f ? -1 : f[x][d];
    }
};
